src/build/cargo.rs

```diff
@@ -88,13 +89,13 @@ pub(super) fn cargo(internals: &Internals) -> BuildResult {
 }
 
 fn run_cargo(
-    compilation_cx: Arc<Mutex<CompilationContext>>,
-    rls_config: Arc<Mutex<Config>>,
+    compilation_cx: &Arc<Mutex<CompilationContext>>,
+    rls_config: &Arc<Mutex<Config>>,
     vfs: Arc<Vfs>,
-    env_lock: Arc<EnvironmentLock>,
+    env_lock: &Arc<EnvironmentLock>,
     compiler_messages: Arc<Mutex<Vec<String>>>,
     analysis: Arc<Mutex<Vec<Analysis>>>,
-    out: Arc<Mutex<Vec<u8>>>,
+    out: &Arc<Mutex<Vec<u8>>>,
```

Comment:
I don't think I agree with these changes - cloning an Arc is cheap and they're already a pointer, so I don't think this change is idiomatic code.

Reply:
I think the existence of both `Rc` and `Arc` shows it _is_ idiomatic to pay a syntactic price caring about avoiding atomic ref counting. Since you mentioned it I had a quick search and this seems correct _(at least https://manishearth.github.io/blog/2015/05/27/wrapper-types-in-rust-choosing-your-guarantees agrees)_

> When sharing data from an Arc in a single thread, it is preferable to share & pointers whenever possible.

Saying that, this isn't really a big deal for Rls so I don't really mind one way or the other. However, since this isn't a false positive, practically it comes down to: Do we disable the clippy lint or not? 

I think the lint is definitely useful for most allocating-on-clone types and disabling above each `Arc` would be too ugly, plus I personally think `&Arc` is fine. So I'd suggest keeping the changes.

But its your call, I'm happy to revert the &s and disable this lint in Rls.

Reply:
While I agree that in general avoiding atomic ref-count increment is good (however same-thread call stack using those will be shallow here, which means that the avoided cost will probably be negligible), I think it's important to convey the ownership here.

I believe, that almost all of these (except `env_lock` and maybe `out` - depending on what `compile_ws` will do to `shell` containing it?) will be sent off-thread inside `Arc::new`'d `RlsExecutor` and so these arguments should be received by value and then just moved to `RlsExecutor`.

(In fact, Cargo's API does not convey that, but `compile_ws` is used in a synchronous and scoped manner, such that passed `Arc`s and `Executor` will not outlive the call here. However, in theory they can so we still have to play nice with Cargo.) 

I know that this is mostly philosophical, but I think it'd be more clear here.

Reply:
I think Manish is wrong there :-) IMO &Box or &Arc is *never* idiomatic (and Clippy lints against the former). If you have `Arc<T>` and can pass by reference as `&T`, that is fine, but if you need the `Arc` bit then you should clone. Cloning an `Arc` is really cheap (in the grand scheme of things), it is certainly more expensive than cloning `Rc`, but that distinction exists for really hot loops in really perf-sensitive code.

Reply:
I've removed the `&Arc` and added `#![allow(needless_pass_by_value)]`.